### PR TITLE
Fix bug where multiple service accounts can't be unpatched at once

### PR DESCRIPTION
**Pull request by a-roberts**
# Changes

Different impl, test code needs to be updated in a follow up PR (see https://github.com/tektoncd/dashboard/issues/681), we need this fix in urgently for 0.2.1.

See the big ol' comment in secrets.js, plenty of manual testing done.

We essentially do a replace patch now with a list of known secrets to remain, per service account per namespace (hence lots of for loops).

Testing done:

**Create a, b, c secrets patched to service account tekton-dashboard**
- Sort the table annotation so c, b, a appears.
- Edit the service account manually so it's b, c, a.
- Delete c and a.

**Next, create five secrets.**
- Assign two secrets to one service account `default` in `default`.
- Assign two secrets to one service account, `tekton-dashboard` in tekton-pipelines.
- Assign another secret to a different service account in `tekton-pipelines`.
- Delete ^^ any in any order.

# Submitter Checklist

These are the criteria that every PR should meet, please check them off as you
review them:

- [ ] Includes [tests](https://github.com/tektoncd/community/blob/master/standards.md#principles) (if functionality changed/added)
- [ ] Includes [docs](https://github.com/tektoncd/community/blob/master/standards.md#principles) (if user facing)
- [x] Commit messages follow [commit message best practices](https://github.com/tektoncd/community/blob/master/standards.md#commit-messages)

_See [the contribution guide](https://github.com/tektoncd/dashboard/blob/master/CONTRIBUTING.md)
for more details._

### CODE REVIEW

**Comment by a-roberts** [comment]
@AlanGreene first go at this, please have a look, admittedly limited testing but better than it is atm definitely

Tested with secrets made against same SA: aaa, bbb, ccc, deleted aaa and ccc and both unpatched OK. Same test but deleted bbb, deleted OK. Same test but deleted aaa, deleted OK.

Gonna repeat that same test now against a couple service accounts just to be sure.

**Comment by a-roberts** [comment]
Fixing tests 😅

**Comment by a-roberts** [comment]
@CarolynMabbott @carlos-logro 

as @AlanGreene's unavailable I'd like to get other reviews on this please. We know there's a bug for sure and this, from our testing, resolves said bug. We'd like it in for the 0.2.1 release I must cut by tomorrow at the latest for the Kabanero folks. Please take a look, test and suggest improvements as soon as you can, thanks!

**Comment by a-roberts** [comment]
Tested this also with

Create three secrets
-> foo (one patched to default then manually to tekton-dashboard)
-> bar (one patched to default only)
-> baz (one patched to tekton-dashboard only)

And all is well, want this merged tonight pleeeeeease @carlos-logro 😄

**Comment by carlos-logro** [comment]
Hi Adam, I'm currently on a Jumpstart Hack-a-thon until 5pm.... that being said, Ill try to review it tonight before I go to bed given the urgency :)

**Comment by a-roberts** [comment]
> code looks good although i did not pull and test. will do it later on but as a first impression the only question I have is to why iterate through the delete promises backwards??

Index preservation, removing from the top of the service account leads to positions changing for remaining secrets on a service account (the secrets patched to it are in array and you must remove by index with the kube API)

**Comment by carlos-logro** [comment]
@a-roberts makes sense! Pulling your branch to test... ill let you know how it went

**Comment by a-roberts** [comment]
New approach implemented, https://github.com/tektoncd/dashboard/issues/681 item made to follow up on tests - confident impl is good, @dibbles and I tested this extensively earlier, as did @CarolynMabbott, and we're no longer suffering from race conditions and index problems.

**Comment by a-roberts** [comment]
/retest

https://tekton-releases.appspot.com/build/tekton-prow/pr-logs/pull/tektoncd_dashboard/678/pull-tekton-dashboard-unit-tests/1189597155866710018/ time out in extension registration spotted

File: src/actions/secrets.js [hunk 1]
```diff
@@ -56,10 +58,72 @@ export function fetchSecrets({ namespace } = {}) {
...
+          // Didn't find one to delete, so update listing of ones to keep
+          if (!found) {
+            const itemAsJson = { name: secret };
+            remainingSecrets.push(itemAsJson);
+          }
+        }
+        // One's been removed so we'll need to now replace the
+        // ServiceAccount's secrets field with the ones to keep
+        if (secretRemoved) {
+          replaceSecretFromServiceAccount(
```
Comment (mnuttall) [inline src/actions/secrets.js]: This method name is just a bit clunky - `patchSecretElementOnServiceAccount` or `updateServiceAccountSecrets` perhaps? Again it's only a minor nit.

File: src/api/index.test.js [hunk 2]
```diff
@@ -17,135 +17,6 @@ import * as index from '.';
 
 beforeEach(jest.resetAllMocks);
 
-const serviceAccountPass = {
```
Comment (mnuttall) [inline src/api/index.test.js]: A lot of tests being deleted here - what's left? What's testing the new unpatch logic?

Comment (a-roberts) [reply src/api/index.test.js]: Yeah, the old approach was flawed and the tests centred around those methods which are now gone - there's nothing testing the new unpatch logic except from manual tests, I discussed this with @CarolynMabbott - what I've requested is for both unit tests and bash tests that would exercise this functionality

We're confident this fix is good but I've not had the time to write the tests owing to our upcoming deadline

File: src/actions/secrets.js [hunk 3]
```diff
@@ -56,10 +58,72 @@ export function fetchSecrets({ namespace } = {}) {
...
+        for (let x = 0; x < allSecrets.length; x += 1) {
+          const secret = allSecrets[x].name;
+          let found = false;
+          // For each secret we want to delete
+          for (let z = 0; z < value.length; z += 1) {
+            const item = value[z];
+            // Name matches?
+            if (item === secret) {
+              found = true;
+              secretRemoved = true;
```
[the remark this answers is not in the document]
Comment (a-roberts) [reply src/actions/secrets.js]: Yeah I'll give this a go

**Comment by a-roberts** [comment]
Many dead clusters later... I've made the changes and done some more basic testing, all still appears to be well 😅

**Pull request merged**